remote/coding/coding.go

```diff
@@ -0,0 +1,334 @@
+package coding
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/drone/drone/model"
+	"github.com/drone/drone/remote"
+	"github.com/drone/drone/remote/coding/internal"
+	"github.com/drone/drone/shared/httputil"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+)
+
+const (
+	defaultURL = "https://coding.net" // Default Coding URL
+)
+
+// Opts defines configuration options.
+type Opts struct {
+	URL        string   // Coding server url.
+	Client     string   // Coding oauth client id.
+	Secret     string   // Coding oauth client secret.
+	Scopes     []string // Coding oauth scopes.
+	Machine    string   // Optional machine name.
+	Username   string   // Optional machine account username.
+	Password   string   // Optional machine account password.
+	SkipVerify bool     // Skip ssl verification.
+}
+
+// New returns a Remote implementation that integrates with a Coding Platform or
+// Coding Enterprise version control hosting provider.
+func New(opts Opts) (remote.Remote, error) {
+	remote := &Coding{
+		URL:        defaultURL,
+		Client:     opts.Client,
+		Secret:     opts.Secret,
+		Scopes:     opts.Scopes,
+		Machine:    opts.Machine,
+		Username:   opts.Username,
+		Password:   opts.Password,
+		SkipVerify: opts.SkipVerify,
+	}
+	if opts.URL != defaultURL {
+		remote.URL = strings.TrimSuffix(opts.URL, "/")
+	}
+
```

Comment:
we should be able to do something like this (which I mentioned in my prior comment) so that we can remove the DRONE_CODING_MACHINE environment variable.

```diff
const (
	defaultURL  = "https://coding.net" // Default Coding URL
+	defaultHost = "coding.net" // Default Coding Host
)


func New(opts Opts) (remote.Remote, error) {
	remote := &Coding{
		URL:        defaultURL,
		Client:     opts.Client,
		Secret:     opts.Secret,
		Scopes:     opts.Scopes,
-		Machine:    opts.Machine,
+		Machine:    defaultHost,
		Username:   opts.Username,
		Password:   opts.Password,
		SkipVerify: opts.SkipVerify,
	}
	if opts.URL != defaultURL {
		remote.URL = strings.TrimSuffix(opts.URL, "/")
+		url, err := url.Parse(remote.URL)
+		if err != nil {
+			return nil, err
+		}
+		host, _, err := net.SplitHostPort(url.Host)
+		if err == nil {
+			url.Host = host
+		}
+		remote.Machine  = host
	}

	// Hack to enable oauth2 access in coding's implementation
	oauth2.RegisterBrokenAuthHeaderProvider(remote.URL)
	return remote, nil
}
```